common/src/main/kotlin/de/hpi/cloud/common/entity/Wrapper.kt

```diff
@@ -0,0 +1,175 @@
+package de.hpi.cloud.common.entity
+
+import de.hpi.cloud.common.Context
+import de.hpi.cloud.common.types.L10n
+import de.hpi.cloud.common.types.LocalDateTime
+import kotlinx.serialization.*
+import kotlinx.serialization.internal.SerialClassDescImpl
+import java.net.URI
+
+@Serializable(with = Wrapper.JsonSerializer::class)
+data class Wrapper<E : Entity<E>>(
+    val type: String,
+    val version: Int = 1,
+    val id: Id<E>,
+    val metadata: Metadata,
+    val value: E
+) {
+    companion object {
+        fun <E : Entity<E>> create(
+            context: Context,
+            companion: Entity.Companion<E>,
+            id: Id<E>,
+            sources: List<L10n<URI>> = emptyList(),
+            permissions: Permissions = emptyMap(),
+            value: E,
+            published: Boolean = true
+        ): Wrapper<E> {
+            return Wrapper(
+                type = companion.type,
+                version = companion.version,
+                id = id,
+                metadata = Metadata(
+                    sources = sources,
+                    permissions = permissions,
+                    events = listOf(CreateEvent.create(context))
+                ),
+                value = value
+            )
+        }
+
+        inline fun <reified E : Entity<E>> jsonSerializerFor(): JsonSerializer<E> =
+            JsonSerializer(E::class.jsonSerializer())
+    }
+
+    @Serializer(forClass = Wrapper::class)
+    class JsonSerializer<E : Entity<E>>(
+        private val entitySerializer: KSerializer<E>
+    ) : KSerializer<Wrapper<E>> {
+        override val descriptor = object : SerialClassDescImpl("Wrapper") {
+            init {
+                addElement("type")
+                addElement("version")
+                addElement("id")
+                addElement("metadata")
+                addElement("value")
+            }
+        }
+
+        override fun serialize(encoder: Encoder, obj: Wrapper<E>) {
+            encoder.beginStructure(descriptor).let {
+                it.encodeStringElement(descriptor, 0, obj.type)
+                it.encodeIntElement(descriptor, 1, obj.version)
+                it.encodeStringElement(descriptor, 2, obj.id.value)
+                it.encodeSerializableElement(descriptor, 3, Metadata.serializer(), obj.metadata)
+                it.encodeSerializableElement(descriptor, 4, entitySerializer, obj.value)
+            }
+        }
+
+        override fun deserialize(decoder: Decoder): Wrapper<E> {
+            val dec = decoder.beginStructure(descriptor)
+            var type: String? = null
+            var version: Int? = null
+            var id: Id<E>? = null
+            var metadata: Metadata? = null
+            var value: E? = null
+
+            loop@ while (true) {
```

Comment:
could be a do-while-loop for better readability instead of label-gotos

Reply:
I'm not sure how using a do-while-loop would remove the label. `break@loop` is required because `break` is a reserved keyword within `when`-expressions.